pkg/storage/imagestore/imagestore.go

```diff
@@ -0,0 +1,2253 @@
+
+						return err
+					}
+				}
+
+				return nil
+			}
+		}
+	}
+
+	if err := is.storeDriver.Delete(blobPath); err != nil {
+		is.log.Error().Err(err).Str("blobPath", blobPath).Msg("unable to remove blob path")
+
+		return err
+	}
+
+	return nil
+}
+
+func (is *ImageStore) garbageCollectIndexReferrers(repo string, index ispec.Index) (bool, error) {
+	var count int
+
+	var err error
+
+	for _, desc := range index.Manifests {
+		switch desc.MediaType {
+		case ispec.MediaTypeImageIndex:
+			indexImage, err := common.GetImageIndex(is, repo, desc.Digest, is.log)
+			if err != nil {
+				is.log.Error().Err(err).Str("repository", repo).Str("digest", desc.Digest.String()).
+					Msg("gc: failed to read multiarch(index) image")
+
+				return false, err
+			}
+
+			gced, err := is.garbageCollectReferrer(repo, index, desc, indexImage.Subject)
+			if err != nil {
+				return false, err
+			}
+
+			// if we gc index then no need to continue searching for referrers inside it.
+			if gced {
+				return true, nil
+			}
+
+			if gced, err = is.garbageCollectIndexReferrers(repo, indexImage); err != nil {
+				return false, err
+			}
+
+			if gced {
+				count++
+			}
+
+		case ispec.MediaTypeImageManifest, artifactspec.MediaTypeArtifactManifest:
+			image, err := common.GetImageManifest(is, repo, desc.Digest, is.log)
+			if err != nil {
+				is.log.Error().Err(err).Str("repo", repo).Str("digest", desc.Digest.String()).
+					Msg("gc: failed to read manifest image")
+
+				return false, err
+			}
+
+			gced, err := is.garbageCollectReferrer(repo, index, desc, image.Subject)
+			if err != nil {
+				return false, err
+			}
+
+			if gced {
+				count++
+			}
+		}
+	}
+
+	return count > 0, err
+}
+
+func (is *ImageStore) garbageCollect(repo string) error {
+	var index ispec.Index
+
+	var err error
+
+	is.log.Info().Msg("gc: manifests with missing referrers")
+
+	// gc all manifests that have a missing subject, stop when no gc happened in a full loop over index.json.
+	stop := false
+	for !stop {
+		index, err = common.GetIndex(is, repo, is.log)
+		if err != nil {
+			return err
+		}
+
+		gced, err := is.garbageCollectIndexReferrers(repo, index)
+		if err != nil {
+			return err
+		}
+
+		// if we delete any manifest then loop again and gc manifests with a subject pointing to this one.
+		stop = !gced
+	}
+
+	is.log.Info().Msg("gc: manifests without tags")
+
+	if err := is.gcUntaggedManifests(index, repo); err != nil {
+		return err
+	}
+
+	is.log.Info().Msg("gc: blobs")
+
+	if err := is.garbageCollectBlobs(is, repo, is.gcDelay, is.log); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (is *ImageStore) garbageCollectReferrer(repo string, index ispec.Index, manifestDesc ispec.Descriptor,
+	subject *ispec.Descriptor,
+) (bool, error) {
+	var gced bool
+
+	var err error
+
+	if subject != nil {
+		// try to find subject in index.json
+		if ok := isManifestReferencedInIndex(index, subject.Digest); !ok {
```

Comment:
I'm not convinced this is enough. Don't we need to look in all possible indexes? Not just the same index where the manifest is found? Or am I misunderstanding?